lib/worker.js

```diff
@@ -330,12 +344,39 @@ class AssetComputeWorker {
         }
     }
 
+    sendEventsBeforeActionTimeout() {
+        return setTimeout(
+            async () => {
+                console.log(`Action is about to timeout in ${timeUntilTimeout()}ms Sending IO events before timeout`);
+                // ensure failure events are sent for any non successful rendition before timeout
+                if(this.renditions){
+                    for (const rendition of this.renditions) {
+                        if (!rendition.eventSent) {
+                            await this.events.sendEvent(EVENT_RENDITION_FAILED, {
+                                rendition: rendition.instructionsForEvent(),
+                                errorReason: Reason.GenericError,
+                                errorMessage: "Timeout",
+                                activationIds: this.params.customWorker ? [process.env.__OW_ACTIVATION_ID] : undefined
+                            });
+                            rendition.eventSent = true;
+                        }
+                    }
+                }
+                // Abort processsing. Process ended abnormally by timeout
+                process.exit(TIMEOUT_EXIT_CODE);
```

Comment:
Killing the process here means the container can't be reused (as it can't accept any requests/activations anymore, even if the current one finishes in time). However the goal is actually the opposite: "avoid having the container killed" (see [NUI-663](https://jira.corp.adobe.com/browse/NUI-663)).

What we should do is see if the rendition callback can be aborted... maybe that needs a new api/spi for workers to implement so they can kill their child processes (etc.).

And then the action should return a result when this timeout handler is called. Returning a result is key for openwhisk to not see this as timeout. Technically this could be done through a promise etc. in the main worker code, which finishes when either the overall processing loop is done (all renditions) or this timeout is called and sets something. A bit tricky but should be possible.

Of course if we cannot abort the current callback, but we return, then this container will get new activations while there might be still some async operations (in nodejs) or a child process churning away. Not sure about the risk here.

Reply:
Ah I see. I misinterpreted the ticket.

> And then the action should return a result when this timeout handler is called. Returning a result is key for openwhisk to not see this as timeout. Technically this could be done through a promise etc. in the main worker code, which finishes when either the overall processing loop is done (all renditions) or this timeout is called and sets something. A bit tricky but should be possible.

^I think we can add that promise inside [here](https://github.com/adobe/asset-compute-sdk/blob/master/lib/worker.js#L108-L123). I will work on that now. 

I will also explore the idea of having an api/spi to abort rendition callbacks.

Reply:
p-cancelable looks good, indeed (just make sure both the cancellation code and the event sending can be sent before the timeout).

Reply:
I mean maybe `process.exit()` is the best and safest solution we can do right now. Or maybe we offer both options, since "canceling" the processing properly is dependent on the worker implementation - we probably cannot guarantee it here.

So one model might be:
1. by default do process.exit() => safe
2. if the worker opts in, support canceling & container reuse

For 2: ideally using an api that automatically opts in, e.g. register an "abort" callback that is supposed to kill any child process etc., and once that is registered we don't do the process.exit.

Would be good to understand how good the cancelable promise approach can work and how much we'd have to refactor for it.

If too difficult, keep this PR to 1., and add 2. later (which is possible if it's opt in).

wdyt?

Reply:
I just refactored the main run loop to use promises so we can gracefully return if the action is about to timeout. The next step is trying/testing out the p-cancelable library to see if we can abort the renditionCB. I will spend some time exploring this. If I find it is unreliable, I will switch back to `process.exit()` for now

Reply:
Ok so I played around with the current behavior and this is what I found. A quick summary of the current behavior:
if the worker is about to timeout (<5s until action timeout), it will send `rendition_failed` events then end the action successfully. It will set `rendition.eventSent = true` right away so there is no race condition in case the rendition finishes before it times out. The same container can be used again.

I tried two different scenarios where the worker timeout is set to 15s:
1. worker sleeps for 30s:
it is not expected here that the worker will complete processing in time before timeout. In testing, it does not complete processing in time. It sends a rendition_failed event for a timeout, does not end up uploading a rendition, and the action finishes successfully. (Same docker image is used in subsequent process calls.)
2. worker sleeps for 10s:
it is expected that the worker will finish before the action actually times out but will be canceled before it finishes processing. The sdk will start "shutting the action down" after 10s, which should be just before the worker finishes processing when you take into account download + set up time.
In testing, it sent a rendition_failed event for timeout but still uploaded the asset in time before the action finished. 


I spent some time trying to figure out how we can "abort processing". The short conclusion is, we can abort processing in all parts of the code that we control (we can add a variable that checks at each step of processing uploading/downloading,etc to see if we went into "timeout mode"). However, we will never be able to control what is inside the custom workers `renditionCallback` and there is no way in javascript (besides process.kill, process.abort, etc.) to kill this cb once it has started since javascript is single-threaded by nature.

So this leaves a few ideas,
1. we keep it as I currently implemented it (describe at the beginning of this comment) and add a boolean variable `this.actionTimedout=true` and check it before each step of the sdk processing.

   My only concern with this is after the action ends due to timeout and returns successfully, what happens in the case 
 that the `renditionCallback` is still running? (first scenario I tried out).  I assume the node process gets killed after the  action returns successfully so there is no `renditionCallback` running in the background of the container...but I cannot confirm this. I do know the action never ends up uploading a rendition and the container gets used again in subsequent calls, which is good.

2. we kill the process when the action is about to timeout (basically I revert my changes made in the last commit). This way we know for sure there are no background processes still running in the container

Reply:
I think you convinced me that 2. `process.exit()` is the safe route. The code also becomes more complex with option 1. looking at the latest commits, which might hurt our maintainability of it going forward (higher risk of new bugs).

Maybe if we see an issue with those self-killed containers in the future we can come back and look at a solution towards option 1.

> <5s until timeout

Maybe make the buffer longer. We might have to send a couple of IO Events and they could fail and retry potentially (say the reason for the worker timing out was slow network etc.). Not sure, maybe 30 seconds? Our shortest worker timeouts are 5 minutes I believe...

Reply:
Alright, I just reverted it back to the older commit (and fixed the previous comments on the review).

As far as timeout buffer, I looked at some new relic data to get some information on how long actions run:
https://insights.newrelic.com/apps/accounts/2045220/asset-compute-production/dashboards/969243?query=SELECT%20median(duration),%20average(duration),%20max(duration)%20FROM%20activation%20WHERE%20namespace%3D%27126848_50549%27%20WHERE%20actionName%20NOT%20IN%20(%27api-process%27,%20%27api-register%27,%20%27core%27,%20%27auth%27)%20since%201%20month%20ago
(median seems to give the 50th percentile as an object and not a number)
```
Action duration over the past month in seconds:
Median: 1.2994146
Avg Duration: 3.55
Max Duration: 3,659
```

I would be hesitant to say 30s buffer time since the default timeout for a runtime action is 60s, and when writing a custom worker (as I observed during the hackathon) most do not change the default value if they don't see a need for it. It would be kind of annoying/confusing when developing to see the action begin to timeout halfway through the time specified. Maybe we do 15 seconds as the timeout buffer?

Reply:
I always love using the Pareto distribution (https://en.wikipedia.org/wiki/Pareto_distribution and/or https://en.wikipedia.org/wiki/Pareto_principle). Meaning: for many events, roughly 80% of the effects come from 20% of the causes.

I would say give 80% of the time to the worker, keep 20% for the timeout handling (means the worker gets 48s out of 60s). Wdyt?

Reply:
Yeah I think that is a good idea.

I would like to calculate the timeout buffer based on max action duration (timeout field in the manifest yaml) but I don't know how we would have access to that besides looking at `process.env.__OW_ACTION_DEADLINE`. I worry the environment variable that is just the time until the action time sout could be off if for some reason the worker does some processing before calling the sdk, etc.

Reply:
However... a full minute feels like a long time to be "timeout handling"...especially since the worker timeout has a grace period of about ~20s.

I went with 15s. Let me know your thoughts. We can always adjust this in the future.

Reply:
ack.

Reply:
@jdelbick regarding median in NRQL: use `percentile(duration, 50)` instead of `median()`, then it works.

Reply:
1. Here is a [query with some 99.x percentiles](https://insights.newrelic.com/apps/accounts/2045220/asset-compute-production/dashboards/969243?query=SELECT%20percentile(duration,%2050)%20AS%20%27median%27,%20percentile(duration,%2099),%20percentile(duration,%2099.9),%20percentile(duration,%2099.94),%20max(duration)%20FROM%20activation%20WHERE%20namespace%3D%27126848_50549%27%20WHERE%20actionName%20LIKE%20%27worker-%25%27%20AND%20actionName%20!%3D%20%27worker-pie-large%27%20AND%20actionName%20!%3D%20%27worker-tika%27%20since%2090%20days%20ago) (and excluding worker-pie-large & worker-tika which are the two I know that have > 5 min timeout).

   Playing around with the percentiles:
   - `99.998%tile` is 264 seconds
   - `99.999%tile` is 296 seconds

   Meaning `< 0.002%` of requests would now fail earlier instead of finishing. (At least I think the query shows _successful_ ones that did not timeout...). Should be ok I guess.

2. How many retries do we get in 15 seconds? If the socket/network timeout is 30 sec, then we would not get any retry in this case. That's why I was thinking about more.

   Default timeout in workers: we could increase this in the generator to say 5 min. At least have a value in the manifest so that it's easy to find and change. But I can also see that Runtime might want us to keep it at 1 min max to avoid too many long running actions where it's not necessary.

   Other than the retry, we only need a fairly fixed amount of time to finish the action (send 1 to few io events), so a general percentage would lead to unneeded long durations for longer timeouts.

   If we need it, we could have a staggered setup: 10 sec for 1 minute, 30 sec for 5 min, 1 min (max?) for 10min+...

   But I am ok with 15 sec for now.

Reply:
Default socket timeout is [30 seconds](https://github.com/adobe/node-fetch-retry#default-behavior) (and I confirmed we use the default value). However, this is [automatically adjusted](https://github.com/adobe/node-fetch-retry/blob/master/index.js#L57-L60) based on `retryMaxDuration` and `retryMaxDuration` is automatically adjusted for [openwhisk action timeout](https://github.com/adobe/node-fetch-retry#apache-openwhisk-action-support)...a bit confusing. 

AFAICS,
`retryMaxDuration` gets adjusted to action deadline if it is greater than action deadline. Then `socketTimeout` will get adjusted to half of the `retryMaxDuration`. So when we send IO events because the action will timeout in ~15s, if my math is correct, it will make the `retryMaxDuration` = ~15s and the the `socketTimeout` = ~7.5s which will result in about 2 retries (maybe more since we know the action won't actually timeout after 15 seconds)

Reply:
Ah, if that's the case, 2 retries sounds good!